### PR TITLE
Add access to Collections in `github/explore` repo

### PR DESCRIPTION
**Pull request by brandonrosage**
This PR adds Collections, using the same workflow and conventions employed on Topics.

### CODE REVIEW

**Comment by brandonrosage** [comment]
I just added a commit that begins to update the docs, including the API, so that it explains the accessibility of Collections within the repo. I'd really appreciate a review from engineers (e.g. @cheshire137 & @talsafran) to make sure I'm using the correct language.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -13,12 +13,12 @@ I'm suggesting these edits to an existing topic:
 Please explain why these changes are necessary:
 
 
-************CURATING A NEW TOPIC PAGE************
+************CURATING A NEW TOPIC OR COLLECTION************
```
Comment (cheshire137) [inline .github/PULL_REQUEST_TEMPLATE.md]: Instead of modifying this to include collections, what about adding a separate `************CURATING A NEW COLLECTION************` below? Then lines 18-20 can be customized for collections specifically, and not try to cover both topics and collections. For example, when curating a new collection, this isn't right: `(e.g. https://github.com/topics/[NAME])`.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -1,39 +1,87 @@
 # Contributing
 
-Hi there! We're excited you've got ideas for an existing or new topic page. Your help makes Explore valuable to others in your community.
+Hi there! We're excited you've got ideas to improve topics and collections. Your helping the community discover valuable information.
```
Comment (cheshire137) [inline CONTRIBUTING.md]: Going from "Your help" to "Your helping" means we should go to "You're". 😄

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -1,39 +1,87 @@
...
-* To update the content, edit the `index.md` inside the topic's folder
+The topic or collection name should match its URL, e.g. `https://github.com/topics/rails` corresponds to the [`topics/rails` directory](https://github.com/github/explore/tree/master/topics/rails).
 
-The [pull request template](./.github/PULL_REQUEST_TEMPLATE.md) provides guidance on the information you need to include.
+To make an improvement, please **open a pull request** with your proposed changes:
 
-**Please fill out the pull request template completely.** If you do not fill out the template, your PR will be closed.
+### Update the image
+
+To update the image, simply replace the image inside the topic or collection's directory
```
Comment (cheshire137) [inline CONTRIBUTING.md]: Needs a `.` at the end.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -1,39 +1,87 @@
...
+
+### Update text & links
```
Comment (cheshire137) [inline CONTRIBUTING.md]: Is an ampersand a special character in Markdown? 🤔  Maybe this should be `&amp;`.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -1,39 +1,87 @@
...
+
+To update text and links, edit the `index.md` inside the topic or collection's directory. These files are formatted using a combination of [front matter](https://jekyllrb.com/docs/frontmatter/) and simple body content.
```
Comment (cheshire137) [inline CONTRIBUTING.md]: I think "front matter" should be capitalized.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -1,39 +1,87 @@
...
 
 Please note that all suggestions must adhere to GitHub's [Community Guidelines](https://help.github.com/articles/github-community-guidelines/) and [Terms of Service](https://help.github.com/articles/github-terms-of-service/). Per our Terms of Service, [you are responsible](https://help.github.com/articles/github-terms-of-service/#d-user-generated-content) for the content you contribute, and you must have the rights to use it.
 
-To propose a new topic page, please **open an pull request** with your proposed changes. The [API docs](./docs/API.md) and [style guide](./docs/styleguide.md) provide guidance on the information you need to include and how it should be formatted.
+To propose a new topic or collection, please **open an pull request** with your proposed additions. The [API docs](./docs/API.md) and [style guide](./docs/styleguide.md) provide guidance on the information you need to include and how it should be formatted.
```
Comment (cheshire137) [inline CONTRIBUTING.md]: `an pull request` => `a pull request`

File: README.md [hunk 7]
```diff
@@ -2,19 +2,17 @@
...
 
-If you want to suggest edits to an existing topic page, or curate a new topic page, read our [contributing guide](CONTRIBUTING.md) to get started. You can [review a list of popular topics that need more context](topics-todo.md) to get an idea of where to start.
+If you want to suggest edits to an existing topic page or collection, or curate a new one, read our [contributing guide](CONTRIBUTING.md) to get started. You can also [review a list of popular topics that need more context](topics-todo.md) to get an idea of where to start.
 
 ## Running tests
 
-There are some lint tests in place to ensure each topic is formatted in the way we expect. Travis
-CI will run the tests automatically. If you want to run the tests yourself locally, you will need
-Ruby and Bundler installed.
+There are some lint tests in place to ensure each topic is formatted in the way we expect. Travis CI will run the tests automatically. If you want to run the tests yourself locally, you will need Ruby and Bundler installed.
```
Comment (cheshire137) [inline README.md]: Should this also mention collections? Maybe not yet if there aren't any automated tests for collection formatting.

File: collections/open-journalism/index.md [hunk 8]
```diff
@@ -0,0 +1,6 @@
+---
+items: fivethirtyeight/data, NYTimes/objective-c-style-guide, newsapps/beeswithmachineguns, voxmedia/meme, propublica/guides, censusreporter/censusreporter, nprapps/app-template, TimeMagazine/babynames, guardian/frontend, dukechronicle/chronline, BloombergMedia/whatiscode, times/cardkit
+display_name: Open journalism
+created_by: benbalter
```
Comment (cheshire137) [inline collections/open-journalism/index.md]: Reading some of these collections, I feel like some automated tests should be added, either in a separate branch that lands first or in this branch, to keep each collection in the right format. Some things I can think of to check:

- `items` list in the right format and contains items that we support, like YouTube URLs, repository `owner/name` strings, and GitHub organizations.
- `display_name`, body, and `items` are present.
- Oxford comma test and similar formatting checks like what we use for topic bodies, but for collection bodies.
- Images are the right size and a valid format.
- `image` field has the same file name as the image in the collection's directory.

File: collections/open-journalism/index.md [hunk 9]
```diff
@@ -0,0 +1,6 @@
+---
+items: fivethirtyeight/data, NYTimes/objective-c-style-guide, newsapps/beeswithmachineguns, voxmedia/meme, propublica/guides, censusreporter/censusreporter, nprapps/app-template, TimeMagazine/babynames, guardian/frontend, dukechronicle/chronline, BloombergMedia/whatiscode, times/cardkit
```
Comment (cheshire137) [inline collections/open-journalism/index.md]: I'm curious why it needs to be a single line. Jekyll Front Matter supports lists with each item on a separate line, e.g.,:

```yaml
items:
- fivethirtyeight/data
- NYTimes/objective-c-style-guide
- newsapps/beeswithmachineguns
- voxmedia/meme
- propublica/guides
- censusreporter/censusreporter
- nprapps/app-template
- TimeMagazine/babynames
- guardian/frontend
- dukechronicle/chronline
- BloombergMedia/whatiscode
- times/cardkit
```

I find this format more readable, and it seems like it'd save some parsing trouble for you. That is, you wouldn't have to split on a comma, the normal Front Matter parser would generate a list for you.

File: docs/API.md [hunk 10]
```diff
@@ -64,3 +64,29 @@ Acceptable formatting:
...
+
+- GitHub repository path (e.g. `defunkt/dotjs`)
+- GitHub username (e.g. `defunkt`)
+- GitHub organization (e.g. `github`)
+- Any web URL (e.g. `https://product.hubspot.com/blog/git-and-github-tutorial-for-beginners`)
+- Any YouTube video URL (e.g. `https://www.youtube.com/watch?v=0fKg7e37bQE`)
+
+A complete `items` key-value pair might look like:
+
+`items: pybee/batavia, Homebrew/brew, https://www.youtube.com/watch?v=dSl_qnWO104`
```
Comment (cheshire137) [inline docs/API.md]: I would encourage allowing a list where each item is on a separate line. I think it makes the Markdown file more readable and I think it would also make it easier to parse when consuming the Markdown file.

File: docs/API.md [hunk 11]
```diff
@@ -64,3 +64,29 @@ Acceptable formatting:
...
+
+### created_by
+(if applicable) GitHub username of the person and/or organization that authored the collection.
```
Comment (cheshire137) [inline docs/API.md]: If we only allow GitHub usernames in this field, an automated test that checks via a regex that the value of `created_by` is a potential username would be good. E.g., `-cats and dogs` is not a valid value.

File: docs/API.md [hunk 12]
```diff
@@ -64,3 +64,29 @@ Acceptable formatting:
...
+
+### display_name
+(required) The collection name that will be displayed on its page (ex. `How to choose (and contribute to) your first open source project`). Should use proper noun capitalization. Emoji are not allowed.
```
Comment (cheshire137) [inline docs/API.md]: > Emoji are not allowed.

😭

File: docs/API.md [hunk 13]
```diff
@@ -64,3 +64,29 @@ Acceptable formatting:
...
+
+### The body of your document
+(required) A longer description of the collection, which will appear on its page. Must be 1,000 characters or less. Some Markdown is allowed, such as links. Emoji are allowed.
```
Comment (cheshire137) [inline docs/API.md]: > Emoji are allowed.

🎉

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 14]
```diff
@@ -1,10 +1,10 @@
...
 I am:
-  - [ ] Suggesting edits to an existing Topic page
-  - [ ] Curating a new Topic page
+  - [ ] Suggesting edits to an existing topic or collection
+  - [ ] Curating a new topic or collection
 
-***********EDITING AN EXISTING TOPIC PAGE************
+***********EDITING AN EXISTING TOPIC OR COLLECTION************
 
 I'm suggesting these edits to an existing topic:
```
[the remark this answers is not in the document]
Comment (talsafran) [reply .github/PULL_REQUEST_TEMPLATE.md]: Add "or collection" here:

> I'm suggesting these edits to an existing topic **or collection**:

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -1,39 +1,87 @@
...
+Note: Updates won't immediately appear once we've merged your PR. We pull in these changes regularly to GitHub.
 
-## Improving an existing topic page
+## Improving an existing topic or collection
 
-If a topic page already exists, it will be listed in the [topics directory](https://github.com/github/explore/tree/master/topics). The topic name should match its URL, e.g. `https://github.com/topics/rails` corresponds to the [`rails` folder](https://github.com/github/explore/tree/master/topics/rails).
+If a topic or collection already exists, it will be listed in its respective directory:
 
-To improve an existing topic page, please **open a pull request** with your proposed changes:
+- [Topics](https://github.com/github/explore/tree/master/topics)
```
Comment (talsafran) [inline CONTRIBUTING.md]: How would you feel about making these appear more directory-ish? Eg:

> If a topic or collection already exists, it will be listed in its respective directory:
> 
> - [`topics/`](https://.com/github/explore/tree/master/topics)
> - [`collections/`](https://github.com/github/explore/tree/master/collections)

File: README.md (hunk 7 above)
Comment (talsafran) [reply README.md]: +1. Shall we add a test in the same vein as the [`test/topics_test.rb`](https://github.com/github/explore/blob/collections/test/topics_test.rb)?

Comment (damaneice) [reply README.md]: I will work on adding a test for collections

File: collections/open-journalism/index.md (hunk 9 above)
Comment (damaneice) [reply collections/open-journalism/index.md]: @cheshire137 @talsafran I will take a look at this

File: docs/API.md (hunk 11 above)
Comment (brandonrosage) [reply docs/API.md]: @damaneice Am I correct in assuming that `created_by` expects a GitHub username? Or can it be any value?

File: docs/API.md (hunk 12 above)
Comment (brandonrosage) [reply docs/API.md]: @talsafran , @damaneice: Should I create an issue to build emoji support in display names in the future?

File: docs/API.md (hunk 11 above)
Comment (damaneice) [reply docs/API.md]: I will add a test to verify that it matches the correct format. But we aren't linking this to an actual GitHub username in the importer.

File: docs/API.md (hunk 12 above)
Comment (damaneice) [reply docs/API.md]: hm.. I would be cool with creating an issue for it.

Comment (talsafran) [reply docs/API.md]: Yah same. Or punt 👢?

Not sure how critical a requirement it is. Fwiw, topics don't currently have emoji support.

File: test/topics_test_helper.rb [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+  Dir["#{topics_dir}/#{topic}/*"].select do |entry|
+    File.file?(entry) && TOPIC_IMAGE_EXTENSIONS.include?(File.extname(entry).downcase)
+  end
+end
+
+def possible_image_file_names_for(topic)
+  TOPIC_IMAGE_EXTENSIONS.map { |ext| "#{topic}#{ext}" }
+end
+
+def metadata_for(topic)
```
Comment (talsafran) [inline test/topics_test_helper.rb]: Some of the methods here are very similar to the ones in the collections helper (other than the directory really). How would you feel about keeping the implementation in one place?

```ruby
# test_helper.rb
def metadata_for(dir, topic_or_collection)
  path = File.join(dir, topic_or_collection, "index.md"
  return unless File.file?(path)

  ...
end
```

**Comment by brandonrosage** [comment]
Looks like @damaneice crushed the test helper for collections _and_ re-formatted collection items to be line-separated. Baller. So is this good to go/approve?

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 17]
```diff
@@ -1,24 +1,34 @@
...
 
 - [ ] I've formatted my changes as a new folder directory, named for the topic as it appears in the URL on GitHub (e.g. `https://github.com/topics/[NAME]`)
 - [ ] My folder contains a `*.png` image (if applicable) and `index.md`
 - [ ] All required fields in my `index.md` conform to the Style Guide and API docs: https://github.com/github/explore/tree/master/docs
 
-Please explain why you think this Topic page should be curated:
+Please explain why you think this topic should be curated:
+
+
+************CURATING A NEW COLLECTION************
```
Comment (cheshire137) [inline .github/PULL_REQUEST_TEMPLATE.md]: Will having the split between curating a topic versus a collection be confusing for contributors? Maybe we should lean on [multiple PR templates](https://github.com/blog/2495-multiple-issue-and-pull-request-templates) and have a separate one for topics versus collections.

File: CONTRIBUTING.md [hunk 18]
```diff
@@ -1,39 +1,87 @@
...
-* To update the content, edit the `index.md` inside the topic's folder
+The topic or collection name should match its URL, e.g. `https://github.com/topics/rails` corresponds to the [`topics/rails` directory](https://github.com/github/explore/tree/master/topics/rails).
 
-The [pull request template](./.github/PULL_REQUEST_TEMPLATE.md) provides guidance on the information you need to include.
+To make an improvement, please **open a pull request** with your proposed changes:
 
-**Please fill out the pull request template completely.** If you do not fill out the template, your PR will be closed.
+### Update the image
+
+To update the image, simply replace the image inside the topic or collection's directory.
```
Comment (cheshire137) [inline CONTRIBUTING.md]: I think `topic` should be possessive here, too: `topic's`

File: CONTRIBUTING.md [hunk 19]
```diff
@@ -1,39 +1,87 @@
...
+url: https://www.algolia.com/
+wikipedia_url: https://en.wikipedia.org/wiki/Algolia
+---
+Algolia is a hosted API for building search into web and mobile applications, with typo-tolerance, fully configurable relevance, and other tools for making great search experiences.
+```
 
-## Curating a new topic page
+---
+
+Similarly, **collections** like "[music](https://raw.githubusercontent.com/github/explore/master/collections/music/index.md)" call out things like their author and display name in key-value pairs -- with a detailed description in the body of the document. Most importantly, though, collections identify their individual collection items in a comma-separated string value for the key "items."
```
Comment (cheshire137) [inline CONTRIBUTING.md]: > collection items in a comma-separated string

Is this still accurate?

File: CONTRIBUTING.md [hunk 20]
```diff
@@ -1,39 +1,87 @@
...
+
+_/collections/music/index.md_:
+
+```
+---
+items: beetbox/beets, scottschiller/SoundManager2, CreateJS/SoundJS, musescore/MuseScore, tomahawk-player/tomahawk, cashmusic/platform, mopidy/mopidy, AudioKit/AudioKit, Soundnode/soundnode-app, gillesdemey/Cumulus, metabrainz/picard, overtone/overtone, samaaron/sonic-pi
```
Comment (cheshire137) [inline CONTRIBUTING.md]: Is this the correct format or are we doing separate line items?

File: CONTRIBUTING.md [hunk 21]
```diff
@@ -1,39 +1,87 @@
...
+display_name: Music
+created_by: jonrohan
+---
+Drop the code bass with these musically themed repositories.
```
Comment (cheshire137) [inline CONTRIBUTING.md]: > Drop the code bass

Get out. 😆

File: CONTRIBUTING.md [hunk 22]
```diff
@@ -1,39 +1,81 @@
...
-**Please fill out the pull request template completely.** If you do not fill out the template, your PR will be closed.
+### Update the image
+
+To update the image, simply replace the image inside the topic or collection's directory.
+
+### Update text and links
+
+To update text and links, edit the `index.md` inside the topic or collection's directory. These files are formatted using a combination of [Front Matter](https://jekyllrb.com/docs/frontmatter/) and simple body content.
+
+For **topics**, you'll notice that, in examples like the topic "[algorithm](https://raw.githubusercontent.com/github/explore/master/topics/algorithm/index.md)," data like the its canonical URL, Wikipedia URL, or display name are called out in key-value pairs; while its detailed description is accounted for in the body of the document.
```
Comment (cheshire137) [inline CONTRIBUTING.md]: > like the its canonical URL

I think you want to drop either 'the' or 'its', but also, the algorithm example doesn't have a `url` field.

File: collections/open-journalism/index.md [hunk 23]
```diff
@@ -0,0 +1,18 @@
+---
+items:
+ - fivethirtyeight/data
+ - NYTimes/objective-c-style-guide
+ - newsapps/beeswithmachineguns
+ - voxmedia/meme
+ - propublica/guides
+ - censusreporter/censusreporter
```
Comment (cheshire137) [inline collections/open-journalism/index.md]: @brandonrosage just trying to pad his line count with all these separate item lines, smh. 😛

File: docs/API.md (hunk 10 above)
Comment (cheshire137) [reply docs/API.md]: I think this line needs updating now that you can put items on separate lines.

File: test/collections_test.rb [hunk 24]
```diff
@@ -0,0 +1,119 @@
...
+      it "has an index.md" do
+        path = File.join(collections_dir, collection, "index.md")
+
+        assert File.file?(path), "expected #{path} to be a file"
+      end
+
+      it "has no unexpected files or directories" do
+        files = Dir["#{collections_dir}/#{collection}/**/*"].reject do |entry|
+          file_name = File.basename(entry)
+          image_files = possible_image_file_names_for_collection(collection)
```
Comment (cheshire137) [inline test/collections_test.rb]: This can be pulled outside of the `reject` loop since it only depends on `collection`.

File: test/collections_test.rb [hunk 25]
```diff
@@ -0,0 +1,119 @@
...
+
+          entry == "." || entry == ".." || file_name == "index.md" ||
+            image_files.include?(file_name)
+        end
+
+        assert_empty files, "expected only index.md and a single image"
```
Comment (cheshire137) [inline test/collections_test.rb]: It looks like the test will complain if an unexpected file comes up, but not if there's more than one image. Maybe rephrase this to `"expected only index.md and valid images"`. A separate test could be added that makes sure there are at most 2 files in the directory.

File: test/collections_test_helper.rb [hunk 26]
```diff
@@ -0,0 +1,51 @@
+# rubocop:disable Metrics/LineLength
+require_relative "./test_helper"
+
+VALID_COLLECTION_METADATA_KEYS = %w[collection created_by display_name image items].freeze
+
+REQUIRED_COLLECTION_METADATA_KEYS = %w[display_name].freeze
```
Comment (cheshire137) [inline test/collections_test_helper.rb]: Seems like `items` and `collection` would also be required for a collection.

File: test/collections_test_helper.rb [hunk 27]
```diff
@@ -0,0 +1,51 @@
...
+
+MAX_COLLECTION_LENGTH = 40
```
Comment (cheshire137) [inline test/collections_test_helper.rb]: Maybe `MAX_COLLECTION_SLUG_LENGTH` so it's clear that's how long the `collection:` key that represents the URL slug can be, as opposed to how many items are allowed in a collection.

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 17 above)
Comment (brandonrosage) [reply .github/PULL_REQUEST_TEMPLATE.md]: ✅ Done

File: test/collections_test_helper.rb (hunk 26 above)
Comment (damaneice) [reply test/collections_test_helper.rb]: I don't believe Collections are required to have items. I will look at adding collection.

Comment (cheshire137) [reply test/collections_test_helper.rb]: What is the point of an empty collection? Seems like at that point, all you have is the name and description, making it essentially a blog post.

Comment (damaneice) [reply test/collections_test_helper.rb]: I am not sure.  @brandonrosage  Should we make items required?

Comment (brandonrosage) [reply test/collections_test_helper.rb]: @damaneice We should make them required. There won't be a usable instance of a Collection without them. They'd just be confusing to users.

**Comment by talsafran** [comment]
:shipit:

**Comment by Boomba71** [comment]
https://allrare4sale.wordpress.com/
